plugin/plugin.go

```diff
@@ -58,57 +61,71 @@ type CLIPlugin struct {
 
 // NewCLIPlugin validate the metadata of the plugin and return a *CLIPlugin.
 func NewCLIPlugin(ctx context.Context, name, path string) (*CLIPlugin, error) {
-	plugin := CLIPlugin{
-		name: name,
-		path: path,
-	}
-
-	// validate metadata
-	metadata, err := plugin.GetMetadata(ctx, &proto.GetMetadataRequest{})
+	// validate file existence
+	fi, err := os.Stat(path)
 	if err != nil {
-		return nil, fmt.Errorf("failed to fetch metadata: %w", err)
+		// Ignore any file which we cannot Stat
+		// (e.g. due to permissions or anything else).
+		return nil, err
 	}
-	if metadata.Name != name {
-		return nil, fmt.Errorf("executable name must be %q instead of %q", binName(metadata.Name), filepath.Base(path))
-	}
-	if err = metadata.Validate(); err != nil {
-		return nil, fmt.Errorf("invalid metadata: %w", err)
+	if !fi.Mode().IsRegular() {
+		// Ignore non-regular files.
+		return nil, ErrNotRegularFile
 	}
 
+	// generate plugin
+	plugin := CLIPlugin{
+		name: name,
+		path: path,
+	}
 	return &plugin, nil
 }
 
 // GetMetadata returns the metadata information of the plugin.
 func (p *CLIPlugin) GetMetadata(ctx context.Context, req *proto.GetMetadataRequest) (*proto.GetMetadataResponse, error) {
-	var resp proto.GetMetadataResponse
-	err := run(ctx, p.name, p.path, req, &resp)
-	return &resp, err
+	var metadata proto.GetMetadataResponse
+	err := run(ctx, p.name, p.path, req, &metadata)
+	if err != nil {
+		return nil, err
+	}
+	// validate metadata
+	if err = validate(&metadata); err != nil {
+		return nil, fmt.Errorf("invalid metadata: %w", err)
+	}
+	if metadata.Name != p.name {
+		return nil, fmt.Errorf("executable name must be %q instead of %q", binName(metadata.Name), filepath.Base(p.path))
+	}
+	return &metadata, nil
 }
 
 // DescribeKey returns the KeySpec of a key.
 func (p *CLIPlugin) DescribeKey(ctx context.Context, req *proto.DescribeKeyRequest) (*proto.DescribeKeyResponse, error) {
 	var resp proto.DescribeKeyResponse
+	req.ContractVersion = proto.ContractVersion
```

Comment:
We should take `ContractVersion` in input as its value depends on the output of `et-plugin-metadata` command.  
Also, updating req struct inside of a method doesnt looks right and reduces readability

Reply:
Is it possible to let the plugin package to decide the ContractVersion and the package user don't need to care about the details? Because the ContractVersion checking is done in the plugin package, the package has enough information to decide the ContractVersion. When we have multiple ContractVersions in the future, we can have a multi-version selector in the package to set the ContractVersion.

Yes, it reduces the readability, so can we updated it like this to give a chance to let the user decide the version?
```go
if req.ContractVersion == ""
    req.ContractVersion = proto.ContractVersion
```

Reply:
So AFAIK contract version is driven by output of `get-plugin-metadata` command. If plugin package can get/store that value and use it for other commands then yes plugin package should full contract version.

Yes, this sounds better
```
if req.ContractVersion == ""
    req.ContractVersion = proto.ContractVersion
```

Reply:
- Added function doc to mention that the ContactVersion can be set automatically.
- We only have one version now, so the `get/store ContractVersion` logic is not required. It should be implemented when we really need it.

Is it acceptable?

Reply:
Yes that sounds good.